src/util/vecdeque.h

```diff
@@ -0,0 +1,261 @@
+            // If not, we need to invoke the destructor for every element separately.
+            while (m_size > size) {
+                std::destroy_at<T>(m_buffer + Index(m_size - 1));
+                --m_size;
+            }
+        }
+    }
+
+public:
+    VecDeque() noexcept = default;
+
+    void resize(size_t size)
+    {
+        if (size < m_size) {
+            // Delegate to ResizeDown when shrinking.
+            ResizeDown(size);
+        } else if (size > m_size) {
+            // When growing, first see if we need to allocate more space.
+            if (size > m_capacity) Reallocate(size);
+            if constexpr (std::is_trivially_constructible_v<T>) {
+                // If T is trivially constructible, just zero-initialize the area being used.
+                size_t first_part = std::min(m_capacity - Index(m_size), size - m_size);
+                if (first_part) std::fill(m_buffer + Index(m_size), m_buffer + Index(m_size) + first_part, 0);
+                if (first_part < size - m_size) std::fill(m_buffer, m_buffer + size - m_size - first_part, 0);
+                m_size = size;
+            } else {
+                // Otherwise invoke the default constructor for every element separately.
+                while (m_size < size) {
+                    std::construct_at<T>(m_buffer + Index(m_size));
+                    ++m_size;
+                }
+            }
+        }
+    }
+
+    void clear() { ResizeDown(0); }
+
+    ~VecDeque()
+    {
+        clear();
+        Reallocate(0);
+    }
+
+    VecDeque& operator=(const VecDeque& other)
+    {
+        clear();
+        Reallocate(other.m_size);
+        if constexpr (std::is_trivially_copyable_v<T>) {
+            size_t first_part = other.FirstPart();
+            if (first_part != 0) {
+                std::memcpy(m_buffer, other.m_buffer + other.m_offset, first_part * sizeof(T));
+            }
+            if (first_part != other.m_size) {
+                std::memcpy(m_buffer + first_part, other.m_buffer, (other.m_size - first_part) * sizeof(T));
+            }
+            m_size = other.m_size;
+        } else {
+            while (m_size < other.m_size) {
+                std::construct_at<T>(m_buffer + Index(m_size), other[m_size]);
+                ++m_size;
+            }
+        }
+        return *this;
+    }
+
+    void swap(VecDeque& other) noexcept
+    {
+        std::swap(m_buffer, other.m_buffer);
+        std::swap(m_offset, other.m_offset);
+        std::swap(m_size, other.m_size);
+        std::swap(m_capacity, other.m_capacity);
+    }
+
+    friend void swap(VecDeque& a, VecDeque& b) noexcept { a.swap(b); }
+    VecDeque& operator=(VecDeque&& other) noexcept { swap(other); return *this; }
+    VecDeque(const VecDeque& other) { *this = other; }
+    VecDeque(VecDeque&& other) noexcept { swap(other); }
+
+    bool friend operator==(const VecDeque& a, const VecDeque& b)
+    {
+        if (a.m_size != b.m_size) return false;
```

Comment:
Just adding for posterity because I went down a rabbit hole for this:

It seems clang's `__is_trivially_equality_comparable` builtin would do what we want here:
> Returns true if comparing two objects of the provided type is known to be equivalent to comparing their object representations. Note that types containing padding bytes are never trivially equality comparable.

Not that it's worth using here.